Sources/PackageRegistry/RegistryClient.swift

```diff
@@ -151,129 +151,181 @@ public final class RegistryClient: Cancellable {
-                    continuation.resume(with: $0)
-                }
-            )
-        }
-    }
-
-    @available(*, noasync, message: "Use the async alternative")
-    public func getPackageMetadata(
-        package: PackageIdentity,
-        timeout: DispatchTimeInterval? = .none,
-        observabilityScope: ObservabilityScope,
-        callbackQueue: DispatchQueue,
-        completion: @escaping (Result<PackageMetadata, Error>) -> Void
-    ) {
-        let completion = self.makeAsync(completion, on: callbackQueue)
-
         guard let registryIdentity = package.registry else {
-            return completion(.failure(RegistryError.invalidPackageIdentity(package)))
+            throw RegistryError.invalidPackageIdentity(package)
         }
 
         guard let registry = self.configuration.registry(for: registryIdentity.scope) else {
-            return completion(.failure(RegistryError.registryNotConfigured(scope: registryIdentity.scope)))
+            throw RegistryError.registryNotConfigured(scope: registryIdentity.scope)
         }
 
-        observabilityScope.emit(debug: "registry for \(package): \(registry)")
-
         let underlying = {
-            self._getPackageMetadata(
+            try await self._getPackageMetadata(
                 registry: registry,
                 package: registryIdentity,
                 timeout: timeout,
                 observabilityScope: observabilityScope,
-                callbackQueue: callbackQueue,
-                completion: completion
+                callbackQueue: callbackQueue
             )
         }
 
         if registry.supportsAvailability {
-            self.withAvailabilityCheck(
+            // The only value that will return is .available, otherwise this check throws an error.
+            if case .available = try await self.withAvailabilityCheck(
                 registry: registry,
                 observabilityScope: observabilityScope,
                 callbackQueue: callbackQueue
-            ) { error in
-                if let error {
-                    return completion(.failure(error))
-                }
-                underlying()
+            ) {
+                return try await underlying()
             }
-        } else {
-            underlying()
         }
+        return try await underlying()
     }
 
-    // marked internal for testing
-    func _getPackageMetadata(
-        registry: Registry,
-        package: PackageIdentity.RegistryIdentity,
-        timeout: DispatchTimeInterval?,
+    @available(*, noasync, message: "Use the async alternative")
+    public func getPackageMetadata(
+        package: PackageIdentity,
+        timeout: DispatchTimeInterval? = .none,
         observabilityScope: ObservabilityScope,
         callbackQueue: DispatchQueue,
         completion: @escaping (Result<PackageMetadata, Error>) -> Void
     ) {
         let completion = self.makeAsync(completion, on: callbackQueue)
+        _ = Task {
```

Comment:
With the unstructured task spawned, how does this handle cancellation?

Reply:
Good point, cancellation is not handled, though was it handled previously? As far as I can tell the `callbackQueue` typically provided is `DispatchQueue.sharedConcurrent` and DispatchQueue doesn't have a nice mechanism for this.

Reply:
`Task.isCancelled` will always be `false` since only structured child tasks get cancellation propagated from parent tasks. `Task` is an unstructured task with no parents, the only primitives for structured concurrency that propagate cancellation are `async let` and `withTaskGroup`.

Reply:
OK, that is consistent with my understanding. Seeing as the sync version of the API didn't handle cancellation before I think its reasonable to keep that behaviour.

However you're correct that the async version _should_ handle cancellation, and I've updated the PR accordingly.